File: input/java/com/floraldroid/input/OperationProtocol.java
/*
 * Copyright 2026 FloralDroid
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.floraldroid.input;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.EOFException;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

final class OperationProtocol {
  static final int MAGIC = 0x46444f31; // FDO1
  static final int VERSION = 1;
  static final int HEADER_SIZE = 24;
  static final int MAX_PAYLOAD_SIZE = 64 * 1024;

  static final int ROUTE_REQUEST = 0x2000;
  static final int ROUTE_RESPONSE = 0x2100;
  static final int ROUTE_EVENT = 0x2200;
  static final int ROUTE_ERROR = 0x2300;

  static final int OP_GENERIC_ERROR = 0x0000;
  static final int OP_BIND_INPUT_TARGET = 0x0100;
  static final int OP_UNBIND_INPUT_TARGET = 0x0101;
  static final int OP_TARGET_INVALIDATED = 0x0102;
  static final int OP_TOUCH = 0x0200;

  static final int ERROR_UNSUPPORTED_MESSAGE = 1;
  static final int ERROR_MALFORMED_MESSAGE = 2;
  static final int ERROR_INVALID_REQUEST_ID = 3;
  static final int ERROR_INTERNAL = 4;

  static final int MODE_EXCLUSIVE = 0;
  static final int ACTION_DOWN = 0;
  static final int ACTION_MOVE = 1;
  static final int ACTION_UP = 2;
  static final int ACTION_CANCEL = 3;

  static final int INVALIDATION_DISPLAY_REMOVED = 1;
  static final int INVALIDATION_GEOMETRY_CHANGED = 2;

  private OperationProtocol() {}

  static final class Header {
    final int operationCode;
    final int routeKind;
    final long requestId;
    final int payloadSize;

    Header(int operationCode, int routeKind, long requestId, int payloadSize) {
      this.operationCode = operationCode;
      this.routeKind = routeKind;
      this.requestId = requestId;
      this.payloadSize = payloadSize;
    }
  }

  static final class BindRequest {
    final int targetSlot;
    final int mode;
    final int displayPort;
    final long streamId;

    BindRequest(int targetSlot, int mode, int displayPort, long streamId) {
      this.targetSlot = targetSlot;
      this.mode = mode;
      this.displayPort = displayPort;
      this.streamId = streamId;
    }
  }

  static final class TouchEvent {
    final int targetSlot;
    final int action;
    final int pointerId;
    final long inputEpoch;
    final long sequence;
    final int x;
    final int y;
    final int pressure;
    final int touchMajor;

    TouchEvent(
        int targetSlot,
        int action,
        int pointerId,
        long inputEpoch,
        long sequence,
        int x,
        int y,
        int pressure,
        int touchMajor) {
      this.targetSlot = targetSlot;
      this.action = action;
      this.pointerId = pointerId;
      this.inputEpoch = inputEpoch;
      this.sequence = sequence;
      this.x = x;
      this.y = y;
      this.pressure = pressure;
      this.touchMajor = touchMajor;
    }
  }

  static Header readHeader(DataInputStream input) throws IOException {
    final int magic;
    try {
      magic = input.readInt();
    } catch (EOFException error) {
      throw error;
    }
    if (magic != MAGIC) {
      throw new IOException("operation packet magic does not match");
    }
    if (input.readUnsignedShort() != VERSION) {
      throw new IOException("operation packet version is unsupported");
    }
    if (input.readUnsignedShort() != HEADER_SIZE) {
      throw new IOException("operation packet header size does not match");
    }
    final int operationCode = input.readUnsignedShort();
    final int routeKind = input.readUnsignedShort();
    final long requestId = Integer.toUnsignedLong(input.readInt());
    final int payloadSize = input.readInt();
    if (input.readInt() != 0) {
      throw new IOException("operation packet reserved field is not zero");
    }
    if (!isKnownRoute(routeKind)) {
      throw new IOException("operation packet route or kind is unsupported");
    }
    if ((routeKind == ROUTE_EVENT) != (requestId == 0)) {
      throw new IOException("operation request id does not match packet kind");
    }
    if (payloadSize < 0 || payloadSize > MAX_PAYLOAD_SIZE) {
      throw new IOException("operation payload exceeds protocol limit");
    }
    return new Header(operationCode, routeKind, requestId, payloadSize);
  }

  static byte[] readPayload(DataInputStream input, int size) throws IOException {
    final byte[] payload = new byte[size];
    input.readFully(payload);
    return payload;
  }

  static void writePacket(
      DataOutputStream output, int operationCode, int routeKind, long requestId, byte[] payload)
      throws IOException {
    if (!isKnownRoute(routeKind)
        || (routeKind == ROUTE_EVENT) != (requestId == 0)
        || requestId < 0
        || requestId > 0xffffffffL
        || payload == null
        || payload.length > MAX_PAYLOAD_SIZE) {
      throw new IOException("invalid outgoing operation packet");
    }
    output.writeInt(MAGIC);
    output.writeShort(VERSION);
    output.writeShort(HEADER_SIZE);
    output.writeShort(operationCode);
    output.writeShort(routeKind);
    output.writeInt((int) requestId);
    output.writeInt(payload.length);
    output.writeInt(0);
    output.write(payload);
    output.flush();
  }

  static BindRequest parseBindRequest(byte[] payload) throws IOException {
    requireSize(payload, 8, "bind input target request");
    if (u8(payload[3]) != 0) {
      throw new IOException("bind input target reserved field is not zero");
    }
    final ByteBuffer buffer = wrap(payload);
    final int targetSlot = u8(payload[0]);
    final int mode = u8(payload[1]);
    final int displayPort = u8(payload[2]);
    final long streamId = Integer.toUnsignedLong(buffer.getInt(4));
    if (mode != MODE_EXCLUSIVE || streamId == 0) {
      throw new IOException("bind input target request is invalid");
    }
    return new BindRequest(targetSlot, mode, displayPort, streamId);
  }

  static int parseUnbindRequest(byte[] payload) throws IOException {
    requireSize(payload, 4, "unbind input target request");
    if (u8(payload[1]) != 0 || u8(payload[2]) != 0 || u8(payload[3]) != 0) {
      throw new IOException("unbind input target reserved field is not zero");
    }
    return u8(payload[0]);
  }

  static TouchEvent parseTouchEvent(byte[] payload) throws IOException {
    requireSize(payload, 24, "touch event");
    final ByteBuffer buffer = wrap(payload);
    if (u8(payload[3]) != 0 || buffer.getInt(20) != 0) {
      throw new IOException("touch event reserved field is not zero");
    }
    final TouchEvent event =
        new TouchEvent(
            u8(payload[0]),
            u8(payload[1]),
            u8(payload[2]),
            Integer.toUnsignedLong(buffer.getInt(4)),
            Integer.toUnsignedLong(buffer.getInt(8)),
            Short.toUnsignedInt(buffer.getShort(12)),
            Short.toUnsignedInt(buffer.getShort(14)),
            Short.toUnsignedInt(buffer.getShort(16)),
            Short.toUnsignedInt(buffer.getShort(18)));
    if (event.action < ACTION_DOWN
        || event.action > ACTION_CANCEL
        || event.pointerId > 31
        || event.inputEpoch == 0
        || event.sequence == 0) {
      throw new IOException("touch event identity is invalid");
    }
    if ((event.action == ACTION_DOWN || event.action == ACTION_MOVE) && event.pressure == 0) {
      throw new IOException("touch down or move has zero pressure");
    }
    if (event.action == ACTION_CANCEL
        && (event.pointerId != 0
            || event.x != 0
            || event.y != 0
            || event.pressure != 0
            || event.touchMajor != 0)) {
      throw new IOException("touch cancel contains pointer data");
    }
    return event;
  }

  static byte[] bindResponse(
      int result,
      int targetSlot,
      int displayPort,
      long streamId,
      long inputEpoch,
      int logicalWidth,
      int logicalHeight,
      int rotation) {
    final ByteBuffer output = allocate(28);
    output.putInt(result);
    output.put((byte) targetSlot);
    output.put((byte) displayPort);
    output.putShort((short) 0);
    output.putInt((int) streamId);
    output.putInt((int) inputEpoch);
    output.putInt(logicalWidth);
    output.putInt(logicalHeight);
    output.putShort((short) rotation);
    output.putShort((short) 0);
    return output.array();
  }

  static byte[] unbindResponse(int result, int targetSlot) {
    final ByteBuffer output = allocate(8);
    output.putInt(result);
    output.put((byte) targetSlot);
    output.put(new byte[3]);
    return output.array();
  }

  static byte[] targetInvalidated(int targetSlot, int reason, long streamId, long inputEpoch) {
    final ByteBuffer output = allocate(12);
    output.put((byte) targetSlot);
    output.put((byte) reason);
    output.putShort((short) 0);
    output.putInt((int) streamId);
    output.putInt((int) inputEpoch);
    return output.array();
  }

  static byte[] errorResponse(int error, int failedOperationCode) {
    final ByteBuffer output = allocate(8);
    output.putInt(error);
    output.putShort((short) failedOperationCode);
    output.putShort((short) 0);
    return output.array();
  }

  private static boolean isKnownRoute(int routeKind) {
    return routeKind == ROUTE_REQUEST
        || routeKind == ROUTE_RESPONSE
        || routeKind == ROUTE_EVENT
        || routeKind == ROUTE_ERROR;
  }

  private static void requireSize(byte[] payload, int expected, String name) throws IOException {
    if (payload == null || payload.length != expected) {
      throw new IOException(name + " payload size is invalid");
    }
  }

  private static ByteBuffer wrap(byte[] payload) {
    return ByteBuffer.wrap(payload).order(ByteOrder.BIG_ENDIAN);
  }

  private static ByteBuffer allocate(int size) {
    return ByteBuffer.allocate(size).order(ByteOrder.BIG_ENDIAN);
  }

  private static int u8(byte value) {
    return Byte.toUnsignedInt(value);
  }
}

File: operation/OperationProtocol.cpp
/*
 * Copyright 2026 FloralDroid
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "floral/device/operation/OperationProtocol.h"

namespace floral::device::operation {
namespace {

constexpr size_t kMagicOffset = 0;
constexpr size_t kVersionOffset = 4;
constexpr size_t kHeaderSizeOffset = 6;
constexpr size_t kOperationCodeOffset = 8;
constexpr size_t kRouteKindOffset = 10;
constexpr size_t kRequestIdOffset = 12;
constexpr size_t kPayloadSizeOffset = 16;
constexpr size_t kReservedOffset = 20;
constexpr size_t kErrorPayloadSize = 8;

bool SetError(std::string* error, const char* message) {
    if (error != nullptr) {
        *error = message;
    }
    return false;
}

void WriteUint16(uint8_t* output, uint16_t value) {
    output[0] = static_cast<uint8_t>(value >> 8);
    output[1] = static_cast<uint8_t>(value);
}

void WriteUint32(uint8_t* output, uint32_t value) {
    for (size_t index = 0; index < sizeof(value); ++index) {
        output[index] = static_cast<uint8_t>(value >> ((sizeof(value) - index - 1) * 8));
    }
}

uint16_t ReadUint16(const uint8_t* input) {
    return static_cast<uint16_t>((static_cast<uint16_t>(input[0]) << 8) | input[1]);
}

uint32_t ReadUint32(const uint8_t* input) {
    uint32_t value = 0;
    for (size_t index = 0; index < sizeof(value); ++index) {
        value = (value << 8) | input[index];
    }
    return value;
}

bool IsKnownPacketKind(uint16_t routeKind) {
    if ((routeKind & kOperationPlaneMask) != kFdo1OperationPlane ||
        (routeKind & kOperationPacketFlagsMask) != 0) {
        return false;
    }
    switch (routeKind & kOperationPacketKindMask) {
        case static_cast<uint16_t>(OperationPacketKind::kRequest) << 8:
        case static_cast<uint16_t>(OperationPacketKind::kResponse) << 8:
        case static_cast<uint16_t>(OperationPacketKind::kEvent) << 8:
        case static_cast<uint16_t>(OperationPacketKind::kErrorResponse) << 8:
            return true;
        default:
            return false;
    }
}

bool HasValidRequestId(uint16_t routeKind, uint32_t requestId) {
    const bool isEvent = IsFdo1RouteKind(routeKind, OperationPacketKind::kEvent);
    return isEvent ? requestId == 0 : requestId != 0;
}

bool IsKnownError(OperationError error) {
    switch (error) {
        case OperationError::kUnsupportedMessage:
        case OperationError::kMalformedMessage:
        case OperationError::kInvalidRequestId:
        case OperationError::kInternalError:
            return true;
    }
    return false;
}

}  // namespace

bool SerializeOperationPacketHeader(const OperationPacketHeader& header,
                                    SerializedOperationPacketHeader* output, std::string* error) {
    if (output == nullptr) {
        return SetError(error, "serialized operation header output is null");
    }
    if (!IsKnownPacketKind(header.route_kind)) {
        return SetError(error, "operation header contains an unsupported route or kind");
    }
    if (!HasValidRequestId(header.route_kind, header.request_id)) {
        return SetError(error, "operation header request id does not match its packet kind");
    }
    if (header.payload_size > kMaximumOperationPayloadSize) {
        return SetError(error, "operation payload exceeds the protocol limit");
    }

    output->fill(0);
    WriteUint32(output->data() + kMagicOffset, kOperationPacketMagic);
    WriteUint16(output->data() + kVersionOffset, kOperationPacketVersion);
    WriteUint16(output->data() + kHeaderSizeOffset,
                static_cast<uint16_t>(kOperationPacketHeaderSize));
    WriteUint16(output->data() + kOperationCodeOffset, header.operation_code);
    WriteUint16(output->data() + kRouteKindOffset, header.route_kind);
    WriteUint32(output->data() + kRequestIdOffset, header.request_id);
    WriteUint32(output->data() + kPayloadSizeOffset, header.payload_size);
    WriteUint32(output->data() + kReservedOffset, 0);
    return true;
}

bool ParseOperationPacketHeader(const SerializedOperationPacketHeader& input,
                                OperationPacketHeader* header, std::string* error) {
    if (header == nullptr) {
        return SetError(error, "parsed operation header output is null");
    }
    if (ReadUint32(input.data() + kMagicOffset) != kOperationPacketMagic) {
        return SetError(error, "operation packet magic does not match");
    }
    if (ReadUint16(input.data() + kVersionOffset) != kOperationPacketVersion) {
        return SetError(error, "operation packet version is unsupported");
    }
    if (ReadUint16(input.data() + kHeaderSizeOffset) != kOperationPacketHeaderSize) {
        return SetError(error, "operation packet header size does not match");
    }
    if (ReadUint32(input.data() + kReservedOffset) != 0) {
        return SetError(error, "operation packet reserved field is not zero");
    }

    OperationPacketHeader parsed;
    parsed.operation_code = ReadUint16(input.data() + kOperationCodeOffset);
    parsed.route_kind = ReadUint16(input.data() + kRouteKindOffset);
    parsed.request_id = ReadUint32(input.data() + kRequestIdOffset);
    parsed.payload_size = ReadUint32(input.data() + kPayloadSizeOffset);
    if (!IsKnownPacketKind(parsed.route_kind)) {
        return SetError(error, "operation header contains an unsupported route or kind");
    }
    if (!HasValidRequestId(parsed.route_kind, parsed.request_id)) {
        return SetError(error, "operation header request id does not match its packet kind");
    }
    if (parsed.payload_size > kMaximumOperationPayloadSize) {
        return SetError(error, "operation payload exceeds the protocol limit");
    }
    *header = parsed;
    return true;
}

bool SerializeOperationErrorPayload(const OperationErrorPayload& source,
                                    std::vector<uint8_t>* payload, std::string* error) {
    if (payload == nullptr) {
        return SetError(error, "operation error payload output is null");
    }
    if (!IsKnownError(source.error)) {
        return SetError(error, "operation error code is unknown");
    }
    payload->assign(kErrorPayloadSize, 0);
    WriteUint32(payload->data(), static_cast<uint32_t>(source.error));
    WriteUint16(payload->data() + 4, source.failed_operation_code);
    return true;
}

bool ParseOperationErrorPayload(const std::vector<uint8_t>& payload, OperationErrorPayload* parsed,
                                std::string* error) {
    if (parsed == nullptr) {
        return SetError(error, "parsed operation error output is null");
    }
    if (payload.size() != kErrorPayloadSize || ReadUint16(payload.data() + 6) != 0) {
        return SetError(error, "operation error payload is invalid");
    }
    OperationErrorPayload result;
    result.error = static_cast<OperationError>(ReadUint32(payload.data()));
    result.failed_operation_code = ReadUint16(payload.data() + 4);
    if (!IsKnownError(result.error)) {
        return SetError(error, "operation error code is unknown");
    }
    *parsed = result;
    return true;
}

}  // namespace floral::device::operation
